community/louvain_directed_multiplex.go

```diff
@@ -0,0 +1,883 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package community
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"sort"
+
+	"golang.org/x/tools/container/intsets"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/graph/internal/ordered"
+)
+
+// DirectedMultiplex is an undirected multiplex graph.
+type DirectedMultiplex interface {
+	Multiplex
+
+	// Layer returns the lth layer of the
+	// multiplex graph.
+	Layer(l int) graph.Directed
+}
+
+// qDirectedMultiplex returns the modularity Q score of the multiplex graph layers
+// subdivided into the given communities at the given resolutions and weights. Q is
+// returned as the vector of weighted Q scores for each layer of the multiplex graph.
+// If communities is nil, the unclustered modularity score is returned.
+// If weights is nil layers are equally weighted, otherwise the length of
+// weights must equal the number of layers. If resolutions is nil, a resolution
+// of 1.0 is used for all layers, otherwise either a single element slice may be used
+// to specify a global resolution, or the length of resolutions must equal the number
+// of layers. The resolution parameter is γ as defined in Reichardt and Bornholdt
+// doi:10.1103/PhysRevE.74.016110.
+// QUndirectedMultiplex will panic if the graph has any layer weight-scaled edge with
```

Comment:
I may go through the comments in a later PR - this one is already very large (I'm not looking forward to the post review clean up).